File: alpha_engine/strategy.py
import math
from datetime import UTC, datetime

from .config import settings
from .db import executable_entry_price
from .models import CriticReport, Market, Opportunity, ProbabilityReport


def hours_left(market: Market) -> float:
    if not market.end_date:
        return math.inf
    end = (
        market.end_date
        if market.end_date.tzinfo
        else market.end_date.replace(tzinfo=UTC)
    )
    return (end - datetime.now(UTC)).total_seconds() / 3600


def eligible(market: Market) -> bool:
    remaining = hours_left(market)
    return (
        market.active
        and not market.closed
        and market.liquidity >= settings.min_liquidity
        and market.volume >= settings.min_volume
        and market.spread <= settings.max_spread
        and settings.min_price <= market.yes_price <= settings.max_price
        and settings.min_hours_to_close <= remaining <= settings.max_hours_to_close
    )


def rank_market(market: Market) -> float:
    liquidity = min(1, math.log10(max(market.liquidity, 1)) / 6)
    volume = min(1, math.log10(max(market.volume, 1)) / 7)
    activity = min(1, math.log10(max(market.volume_24h, 1)) / 5)
    spread = max(0, 1 - market.spread / max(settings.max_spread, 0.001))

    remaining = hours_left(market)
    if math.isinf(remaining):
        horizon = 0.0
    else:
        horizon = max(
            0.0,
            1 - remaining / max(settings.max_hours_to_close, 1.0),
        )

    market.rank_score = 100 * (
        0.30 * liquidity
        + 0.25 * volume
        + 0.20 * activity
        + 0.15 * spread
        + 0.10 * horizon
    )
    return market.rank_score


def kelly_binary(probability: float, price: float) -> float:
    if price <= 0 or price >= 1:
        return 0
    odds = (1 - price) / price
    return max(0, (odds * probability - (1 - probability)) / odds)


def build_opportunity(
    market: Market,
    probability: ProbabilityReport,
    critic: CriticReport,
    current_exposure: float,
) -> Opportunity:
    side = "YES" if probability.probability_yes >= market.yes_price else "NO"
    fair = (
        probability.probability_yes
        if side == "YES"
        else 1 - probability.probability_yes
    )
    executable_price = executable_entry_price(market, side)
    gross_edge = fair - executable_price
    net_edge = gross_edge - settings.slippage_buffer
    kelly = kelly_binary(fair, executable_price) * settings.kelly_fraction
    max_stake = settings.bankroll_usdc * settings.max_position_pct
    exposure_room = max(
        0,
        settings.bankroll_usdc * settings.max_total_exposure_pct
        - current_exposure,
    )
    stake = min(settings.bankroll_usdc * kelly, max_stake, exposure_room)
    expected_value = (
        (
            fair * (1 - executable_price)
            - (1 - fair) * executable_price
        )
        / executable_price
        if executable_price > 0
        else -1
    )

    decision = "OPEN"
    reason = "edge and risk thresholds passed"
    if (
        net_edge < settings.min_net_edge
        or probability.confidence < settings.min_confidence
        or critic.risk_score > settings.max_critic_risk
        or critic.recommendation == "REJECT"
        or stake < 1
    ):
        decision = "REJECT"
        reason = "threshold or portfolio constraint failed"
    elif critic.recommendation == "REDUCE":
        stake *= 0.5
        decision = "WATCH"
        reason = "critic requested reduced conviction"

    score = (
        100
        * max(0, net_edge)
        * probability.confidence
        * (1 - critic.risk_score)
    )
    return Opportunity(
        market=market,
        side=side,
        fair_probability=fair,
        market_probability=executable_price,
        gross_edge=gross_edge,
        net_edge=net_edge,
        expected_value_per_dollar=expected_value,
        confidence=probability.confidence,
        critic_risk=critic.risk_score,
        kelly_fraction=kelly,
        stake_usdc=round(stake, 2),
        score=score,
        decision=decision,
        reason=reason,
        probability_report=probability,
        critic_report=critic,
    )

File: worker/src/strategy.ts
import type { Settings } from "./config";
import type { CriticReport, Decision, Market, Opportunity, ProbabilityReport } from "./types";
import { executableEntryPrice } from "./db";

export function hoursLeft(market: Market): number {
  if (!market.endDate) return Infinity;
  const end = new Date(market.endDate).getTime();
  return (end - Date.now()) / 3600000;
}

export function eligible(market: Market, settings: Settings): boolean {
  const remaining = hoursLeft(market);
  return (
    market.active &&
    !market.closed &&
    market.liquidity >= settings.minLiquidity &&
    market.volume >= settings.minVolume &&
    market.spread <= settings.maxSpread &&
    settings.minPrice <= market.yesPrice &&
    market.yesPrice <= settings.maxPrice &&
    settings.minHoursToClose <= remaining &&
    remaining <= settings.maxHoursToClose
  );
}

export function rankMarket(market: Market, settings: Settings): number {
  const liquidity = Math.min(1, Math.log10(Math.max(market.liquidity, 1)) / 6);
  const volume = Math.min(1, Math.log10(Math.max(market.volume, 1)) / 7);
  const activity = Math.min(1, Math.log10(Math.max(market.volume24h, 1)) / 5);
  const spreadScore = Math.max(0, 1 - market.spread / Math.max(settings.maxSpread, 0.001));

  const remaining = hoursLeft(market);
  const horizon = !isFinite(remaining)
    ? 0
    : Math.max(0, 1 - remaining / Math.max(settings.maxHoursToClose, 1));

  market.rankScore =
    100 * (0.3 * liquidity + 0.25 * volume + 0.2 * activity + 0.2 * spreadScore + 0.05 * horizon);
  return market.rankScore;
}

export function kellyBinary(probability: number, price: number): number {
  if (price <= 0 || price >= 1) return 0;
  const odds = (1 - price) / price;
  return Math.max(0, (odds * probability - (1 - probability)) / odds);
}

export function requiredEdge(market: Market, settings: Settings): number {
  const estimatedRoundTripCost =
    market.spread + 2 * settings.slippageBuffer + 2 * settings.paperFeePct;
  return Math.max(settings.minNetEdge, estimatedRoundTripCost * settings.minEdgeCostMultiple);
}

export function buildOpportunity(
  market: Market,
  probability: ProbabilityReport,
  critic: CriticReport,
  currentExposure: number,
  settings: Settings,
): Opportunity {
  const side = probability.probability_yes >= market.yesPrice ? "YES" : "NO";
  const fair = side === "YES" ? probability.probability_yes : 1 - probability.probability_yes;
  const executablePrice = executableEntryPrice(market, side);
  const grossEdge = fair - executablePrice;
  const netEdge = grossEdge - settings.slippageBuffer;
  const threshold = requiredEdge(market, settings);

  const kelly = kellyBinary(fair, executablePrice) * settings.kellyFraction;
  const maxStake = settings.bankrollUsdc * settings.maxPositionPct;
  const exposureRoom = Math.max(
    0,
    settings.bankrollUsdc * settings.maxTotalExposurePct - currentExposure,
  );
  const stake = Math.min(settings.bankrollUsdc * kelly, maxStake, exposureRoom);
  const expectedValue =
    executablePrice > 0
      ? (fair * (1 - executablePrice) - (1 - fair) * executablePrice) / executablePrice
      : -1;

  const evidenceSources = new Set(probability.source_urls).size;
  let decision: Decision = "OPEN";
  let reason = "V6 cost-aware thresholds passed";

  if (evidenceSources < settings.minEvidenceSources) {
    decision = "REJECT";
    reason = "insufficient independent evidence";
  } else if (netEdge < threshold) {
    decision = "REJECT";
    reason = `net edge ${netEdge.toFixed(3)} below cost-aware threshold ${threshold.toFixed(3)}`;
  } else if (probability.confidence < settings.minConfidence) {
    decision = "REJECT";
    reason = "forecast confidence below V6 threshold";
  } else if (critic.risk_score > settings.maxCriticRisk) {
    decision = "REJECT";
    reason = "critic risk above V6 threshold";
  } else if (critic.recommendation !== "ACCEPT") {
    decision = "REJECT";
    reason = `critic recommendation ${critic.recommendation}`;
  } else if (stake < 1) {
    decision = "REJECT";
    reason = "insufficient portfolio room";
  }

  const score = 100 * Math.max(0, netEdge - threshold) * probability.confidence * (1 - critic.risk_score);

  return {
    market,
    side,
    fairProbability: fair,
    marketProbability: executablePrice,
    grossEdge,
    netEdge,
    expectedValuePerDollar: expectedValue,
    confidence: probability.confidence,
    criticRisk: critic.risk_score,
    kellyFraction: kelly,
    stakeUsdc: Math.round(stake * 100) / 100,
    score,
    decision,
    reason,
    probabilityReport: probability,
    criticReport: critic,
  };
}
